### PR TITLE
Manage route controllers in separate ns

### PR DESCRIPTION
**Pull request by atiratree**
following up on https://github.com/openshift/openshift-controller-manager/pull/234 

TODO:
- [ ] #256 

4.11 backports that introduce leases
- [x] https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/251 
- [x] https://github.com/openshift/openshift-controller-manager/pull/232

### CODE REVIEW

File: pkg/operator/sync_openshiftcontrollermanager_v311_00.go [hunk 1]
```diff
@@ -33,75 +33,128 @@ import (
...
-	if actualDaemonSet == nil {
+	if actualDaemonSet == nil || actualRCDeployment == nil {
 		return syncReturn(c, errors, originalOperatorConfig, operatorConfig)
 	}
+	available := actualDaemonSet.Status.NumberAvailable > 0
+	rcAvailable := actualRCDeployment.Status.AvailableReplicas > 0
 
 	// manage status
-	if actualDaemonSet.Status.NumberAvailable > 0 {
+	if available && rcAvailable {
```
Comment (atiratree) [inline pkg/operator/sync_openshiftcontrollermanager_v311_00.go:117]: sharing the conditions so both need to be available to flag available as true

File: bindata/v3.11.0/openshift-controller-manager/route-controller-deploy.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        ports:
+          - containerPort: 8443
+        terminationMessagePolicy: FallbackToLogsOnError
+        volumeMounts:
+        - mountPath: /var/run/configmaps/config
+          name: config
+        - mountPath: /var/run/configmaps/client-ca
+          name: client-ca
+        - mountPath: /var/run/secrets/serving-cert
+          name: serving-cert
```
[the remark this answers is not in the document]
Comment (deads2k) [reply bindata/v3.11.0/openshift-controller-manager/route-controller-deploy.yaml:60]: > this is missing the proxy-ca-bundles and proxy envs compared to the original DS as I have not seen/found a use case for these

acceptable.

File: bindata/v3.11.0/openshift-controller-manager/route-controller-deploy.yaml [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        livenessProbe:
+          initialDelaySeconds: 30
+          httpGet:
+            scheme: HTTPS
+            port: 8443
+            path: healthz
+        readinessProbe:
+          failureThreshold: 10
```
Comment (deads2k) [inline bindata/v3.11.0/openshift-controller-manager/route-controller-deploy.yaml:68]: we should come back to this post-merge.  it looks too long.

File: pkg/operator/operator.go [hunk 4]
```diff
@@ -52,18 +67,22 @@ func NewOpenShiftControllerManagerOperator(
...
-		rateLimiter:          flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
-		recorder:             recorder,
+		targetImagePullSpec:       targetImagePullSpec,
+		operatorConfigClient:      operatorConfigClient,
+		countNodes:                countNodes,
+		ensureAtMostOnePodPerNode: ensureAtMostOnePodPerNode,
+		proxyLister:               proxyInformer.Lister(),
+		kubeClient:                kubeClient,
+		configMapsGetter:          v1helpers.CachedConfigMapGetter(kubeClient.CoreV1(), kubeInformers),
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "KubeApiserverOperator"),
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/operator/operator.go:83]: > should I open additional PR for this?

yes

File: pkg/operator/sync_openshiftcontrollermanager_v311_00.go [hunk 5]
```diff
@@ -33,75 +33,128 @@ import (
...
 	_, serviceCAModified, err := manageOpenShiftServiceCAConfigMap_v311_00_to_latest(c.kubeClient, c.configMapsGetter, c.recorder)
 	if err != nil {
-		errors = append(errors, fmt.Errorf("%q: %v", "openshift-service-ca", err))
+		errors = append(errors, fmt.Errorf("%q %q: %v", operandName, "openshift-service-ca", err))
 	}
 
 	_, globalCAModified, err := manageOpenShiftGlobalCAConfigMap_v311_00_to_latest(c.kubeClient, c.configMapsGetter, c.recorder)
 	if err != nil {
-		errors = append(errors, fmt.Errorf("%q: %v", "openshift-global-ca", err))
+		errors = append(errors, fmt.Errorf("%q %q: %v", operandName, "openshift-global-ca", err))
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/operator/sync_openshiftcontrollermanager_v311_00.go:71]: This should be refactored later.  We can discuss at the control plane arch call next week.

File: pkg/operator/sync_openshiftcontrollermanager_v311_00.go [hunk 6]
```diff
@@ -180,6 +242,16 @@ func manageOpenShiftControllerManagerClientCA_v311_00_to_latest(client coreclien
 	return caChanged, nil
 }
 
+func manageRouteControllerManagerClientCA_v311_00_to_latest(client coreclientv1.ConfigMapsGetter, recorder events.Recorder) (bool, error) {
```
Comment (deads2k) [inline pkg/operator/sync_openshiftcontrollermanager_v311_00.go:245]: this smells weird.  I think we should cleanup in the future.

**Comment by deads2k** [comment]
the overall approach seems ok.  I'll leave the detailed review to @ingvagabund 

/approve

/assign @ingvagabund

File: pkg/operator/operator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (atiratree) [reply pkg/operator/operator.go:83]: done: https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/257

File: manifests/0000_90_openshift-controller-manager-operator_04_operand-route-controller-servicemonitor.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
```
Comment (ingvagabund) [inline manifests/0000_90_openshift-controller-manager-operator_04_operand-route-controller-servicemonitor.yaml:2]: This role is already defined in bindata/v3.11.0/openshift-controller-manager/route-controller-servicemonitor-role.yaml in https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/255/commits/d505b981d68a79a02a0038dba60e40f6647b9a5e#diff-ddf2d7b29e1f246cc23d33358c5891d71becce69bd264c2495d5f8f607a2d3a6

File: manifests/0000_90_openshift-controller-manager-operator_04_operand-route-controller-servicemonitor.yaml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  - services
+  - endpoints
+  - pods
+  verbs:
+  - get
+  - list
+  - watch
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
```
Comment (ingvagabund) [inline manifests/0000_90_openshift-controller-manager-operator_04_operand-route-controller-servicemonitor.yaml:22]: This role binding is already defined in bindata/v3.11.0/openshift-controller-manager/route-controller-servicemonitor-rolebinding.yaml in https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/255/commits/d505b981d68a79a02a0038dba60e40f6647b9a5e#diff-6683cf67175abdad3fc07c757a5fe0e2c62b0c88f642989b1873bcd4f3f224a0

File: pkg/operator/starter.go [hunk 9]
```diff
@@ -153,7 +154,20 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
...
+			"v3.11.0/openshift-controller-manager/route-controller-informer-clusterrole.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-informer-clusterrolebinding.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-leader-role.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-leader-rolebinding.yaml",
 			"v3.11.0/openshift-controller-manager/route-controller-ns.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-sa.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-separate-sa-role.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-separate-sa-rolebinding.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-servicemonitor-role.yaml",
+			"v3.11.0/openshift-controller-manager/route-controller-servicemonitor-rolebinding.yaml",
```
Comment (ingvagabund) [inline pkg/operator/starter.go:170]: Both service monitor role and rolebinding are defined in `manifests/0000_90_openshift-controller-manager-operator_04_operand-route-controller-servicemonitor.yaml`. Will not CVO manage both role and rolebinding as well?

File: pkg/operator/sync_openshiftcontrollermanager_v311_00.go [hunk 10]
```diff
@@ -207,6 +279,32 @@ func manageOpenShiftControllerManagerConfigMap_v311_00_to_latest(kubeClient kube
 	return resourceapply.ApplyConfigMap(context.Background(), client, recorder, requiredConfigMap)
 }
 
+// similar logic for route-controller-manager in manageOpenShiftControllerManagerConfigMap_v311_00_to_latest
+func manageRouteControllerManagerConfigMap_v311_00_to_latest(kubeClient kubernetes.Interface, client coreclientv1.ConfigMapsGetter, recorder events.Recorder, operatorConfig *operatorapiv1.OpenShiftControllerManager) (*corev1.ConfigMap, bool, error) {
+	configMap := resourceread.ReadConfigMapV1OrDie(v311_00_assets.MustAsset("v3.11.0/openshift-controller-manager/route-controller-cm.yaml"))
+	defaultConfig := v311_00_assets.MustAsset("v3.11.0/config/defaultconfig.yaml")
```
[the remark this answers is not in the document]
Comment (atiratree) [reply pkg/operator/sync_openshiftcontrollermanager_v311_00.go:285]: We use the same approach and share same config as for OCM (even after all the changes). So user can use this and disabled/enable the controllers via the overrides if they want.

Do we want to remove this functionality and start both controllers `"openshift.io/ingress-ip"` and `"openshift.io/ingress-to-route"` hardcoded every time?

[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/operator/sync_openshiftcontrollermanager_v311_00.go:285]: I see now. Sharing the config file across both OCM and the route controller manager. No objection here then.

File: pkg/operator/starter.go (hunk 9 above)
Comment (ingvagabund) [reply pkg/operator/starter.go:170]: Checking https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/manifests/0000_90_openshift-controller-manager-operator_03_operand-servicemonitor.yaml:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: prometheus-k8s
  namespace: openshift-controller-manager
  annotations:
    include.release.openshift.io/self-managed-high-availability: "true"
    include.release.openshift.io/single-node-developer: "true"
...
```
vs. https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/bindata/v3.11.0/openshift-controller-manager/servicemonitor-role.yaml:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: prometheus-k8s
  namespace: openshift-controller-manager
```

The former is missing the annotations. We need to revisit the manifests later to see whether we can remove the latter to remove duplicates.

Comment (atiratree) [reply pkg/operator/starter.go:170]: ack, let's revisit this later

**Pull request merged**